crates/sui-graphql-client/src/lib.rs

```diff
@@ -0,0 +1,584 @@
+    }
+
+    pub fn data(&self) -> &[T] {
+        &self.data
+    }
+
+    fn new(page_info: PageInfo, data: Vec<T>) -> Self {
+        Self { page_info, data }
+    }
+}
+
+/// The Client is a GraphQL client for interacting with the Sui blockchain.
+/// By default, it uses Reqwest as the HTTP client.
+pub struct Client {
+    /// The URL of the GraphQL server.
+    rpc: String,
+    /// GraphQL supports multiple versions of the service. By default, the client uses the stable
+    /// version of the service.
+    rpc_version: Option<String>,
+    inner: reqwest::Client,
+}
+
+impl Client {
+    // ===========================================================================
+    // Client Misc API
+    // ===========================================================================
+
+    /// Create a new GraphQL client with the provided server address.
+    pub fn new(server: &str) -> Result<Self, Error> {
+        reqwest::Url::parse(server).map_err(|_| anyhow!("Invalid URL: {}", server))?;
+
+        let client = Client {
+            rpc: server.to_string(),
+            rpc_version: None,
+            inner: reqwest::Client::new(),
+        };
+        Ok(client)
+    }
+
+    /// Create a new GraphQL client connected to the `mainnet` GraphQL server: {MAINNET_HOST}.
+    pub fn new_mainnet() -> Result<Self, Error> {
+        Self::new(MAINNET_HOST)
+    }
+
+    /// Create a new GraphQL client connected to the `testnet` GraphQL server: {TESTNET_HOST}.
+    pub fn new_testnet() -> Result<Self, Error> {
+        Self::new(TESTNET_HOST)
+    }
+
+    /// Create a new GraphQL client connected to the `devnet` GraphQL server: {DEVNET_HOST}.
+    pub fn new_devnet() -> Result<Self, Error> {
+        Self::new(DEVNET_HOST)
+    }
+
+    /// Create a new GraphQL client connected to the `localhost` GraphQL server:
+    /// {DEFAULT_LOCAL_HOST}.
+    pub fn new_localhost(&mut self) -> Result<Self, Error> {
+        Self::new(DEFAULT_LOCAL_HOST)
+    }
+
+    /// Set the version for the GraphQL GraphQL client. The default set version is stable.
+    ///
+    /// Use the `service_config` API to get the `available_versions` field to know which versions
+    /// are supported by this service.
+    pub fn set_version(&mut self, version: Option<&str>) -> &mut Self {
+        self.rpc_version = version.map(|v| v.to_string());
+        self
+    }
+
+    /// Set the server address for the GraphQL GraphQL client. It should be a valid URL with a host and
+    /// optionally a port number.
+    pub fn set_rpc_server(&mut self, server: &str) -> Result<&mut Self, Error> {
+        reqwest::Url::parse(server)?;
+        self.rpc = server.to_string();
+        Ok(self)
+    }
+
+    /// Return the URL for the GraphQL GraphQL client after checking if the version is set.
+    fn url(&self) -> String {
+        if let Some(version) = &self.rpc_version {
+            format!("{}/{}", self.rpc, version)
+        } else {
+            self.rpc.to_string()
+        }
+    }
```

Comment:
I would return a borrowed reference to the url vs allocating a string each time here. if the version is a required part of the URL, then it should always be a part of the url that is being used